### PR TITLE
Fix #3121: use Right as default driving side when driving side is unavailable

### LINKED ISSUE #3121: InstructionView requires route progress before banner instructions
When interacting with the `InstructionView`, it requires the `RouteProgress` value to be provided through
https://github.com/mapbox/mapbox-navigation-android/blob/058459d9336a2bd454394ba48732d5e4eb5af921/libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java#L323

before we can provide banner instructions through
https://github.com/mapbox/mapbox-navigation-android/blob/058459d9336a2bd454394ba48732d5e4eb5af921/libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java#L330

If not, the `currentStep` is not initialized and we run into:
```
Process: com.mapbox.onetap.app, PID: 4246
java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String com.mapbox.api.directions.v5.models.LegStep.drivingSide()' on a null object reference
    at com.mapbox.navigation.ui.instruction.InstructionView.updateBannerInstructionsWith(InstructionView.java:330)
```

We should preferably remove this dependency, or reshape the `update*` APIs so a developer cannot make this mistake.

/cc @abhishek1508 @JunDai

### PR DESCRIPTION
**Pull request by JunDai**
## Description

This is a proposal to fix #3121: use Right as default driving side when driving side is unavailable.
Also add an API to allow user set the default "driving side".

- [x] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [ ] I have added the appropriate milestone and project boards

### Goal

To make InstructionView API crash free.

### Implementation

Add an API to let the developer to set the default driving side, this value has higher priority than the one from RouteProgress. If developer doesn't set a default driving side and RouteProgress isn't available, then will use `Right` as default driving side. So we could avoid NPE crash.

## Testing

Please describe the manual tests that you ran to verify your changes

- [x] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [x] I have tested via a test drive, or a simulation/mock location app
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR
- [ ] We might need to update / push `api/current.txt` files after running `$> make 1.0-core-update-api` (Core) / `$> make 1.0-ui-update-api` (UI) if there are changes / errors we're 🆗 with (e.g. `AddedMethod` changes are marked as errors but don't break SemVer) 🚀 If there are SemVer breaking changes add the `SEMVER` label. See [Metalava](https://github.com/mapbox/mapbox-navigation-android/blob/master/docs/metalava.md) docs

### CODE REVIEW

**Comment by JunDai** [comment]
@LukasPaczos / @Guardiola31337 / @cafesilencio / @abhishek1508 : this is a proposal to fix #3121 
Feel free to discuss pros and cons, so we can fix it asap.

thanks all 🙏

**Comment by LukasPaczos** [comment]
> Add an API to let the developer to set the default driving side, this value has higher priority than the one from RouteProgress.

I'm afraid this might be pushing too much responsibility to the user. Nav Native/Directions API should always be able to correctly determine the driving side. In this case, after wanting to set the fallback, the user would have to always control the driving side themselves in different geographies making the usage more complicated in the end.

Instead of exposing an additional field and the setter, how about we add an overload:
```
public void updateBannerInstructionsWith(BannerInstructions instructions, String drivingSideFallback) {
  if (instructions != null) {
    String drivingSide = ManeuverModifier.RIGHT;
    if (currentStep != null) {
      drivingSide = currentStep.drivingSide();
    } else if (drivingSideFallback != null && !drivingSideFallback.isEmpty()) {
      drivingSide = drivingSideFallback;
    }
    updateBannerInstructions(instructions.primary(),
        instructions.secondary(), instructions.sub(), drivingSide);
  }
}
```

**Comment by JunDai** [comment]
@LukasPaczos - thanks for looking at the code.

I think the idea is same to get a default driving side before it's available. To do this, we have to assign a default one, `RIGHT` side driving, and we also need to consider the case that user want to set the default driving side to `LEFT`.

I don't think my change adds responsibility to the developer as they only need to call the `setter` when they want a default `LEFT` driving side. If that's not set, then we still use the `directionAPI`'s return value first. And the user doesn't need to change their implementation because the `updateBannerInstructionsWith` doesn't change. Also the API will be clear since it has its own responsibility.  -- _we can discuss this_

I would say the only difference between our change is the priority between user's default setting and `directionAPI`'s return value. This one we can also discuss and get an agreement should be fine. You mentioned `Nav Native/Directions API should always be able to correctly determine the driving side.`, I am happy to use that our API's value first. Something like this:
```java
  private String getDrivingSide() {
    if (currentStep != null) {
      return currentStep.drivingSide();
    } else if (drivingSide!= null) {
      return drivingSide;
    } else {
      return ManeuverModifier.RIGHT;
    }
  }

  public void setDefaultDrivingSide(boolean isLeftDriving) {
    if (isLeftDriving) {
      drivingSide = ManeuverModifier.LEFT;
    } else {
      drivingSide = ManeuverModifier.RIGHT;
    }
  }

```

As I said, this is just a proposal/draft to fix the issue. 

cc: @Guardiola31337 / @cafesilencio / @abhishek1508

**Comment by JunDai** [comment]
@LukasPaczos - I updated the `getDrivingSide()` method so the `RouteProgress`'s return value has higher priority than user set value.

To summarize, if user is in a right driving localization, user doesn't need to make any change of their work. If user is in a left driving location, they will need to call `setDrivingSide(true)` after create the InstructionView. So later when we get the `RouteProgress`, the UX won't suddenly change from right driving UX to left driving.

Given that now we are close to Jun 30 release, this temp fix should be enough and low risk.

**Comment by LukasPaczos** [comment]
The current proposal works just fine but introduces a new public method that again requires the correct method invocation sequence to take any effect and forces us to maintain another nullable field. Also, the field becomes virtually obsolete if route progress is available which can be another point of confusion.

I'd still suggest considering below overload pattern as a temp fix to make the APIs easier to use:
```
public void updateBannerInstructionsWith(BannerInstructions instructions)
public void updateBannerInstructionsWith(BannerInstructions instructions, String drivingSideFallback)
```

**Comment by JunDai** [comment]
@LukasPaczos - changes are made accordingly. Please take a look 👍

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java [hunk 1]
```diff
@@ -117,6 +118,7 @@
   private Animation rerouteSlideUpTop;
   private Animation rerouteSlideDownTop;
   private LegStep currentStep;
+  private String drivingSide = null;
```
Comment (LukasPaczos) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java:121]: This is unsued, isn't it?

**Pull request merged**